File: src/hillclimber/monitor.ts
/**
 * HillClimber Monitor Module
 *
 * Validates actions BEFORE execution to catch obvious mistakes.
 * Rule-based, no LLM needed for fast validation.
 *
 * Part of the MAP-inspired architecture for 10x better HillClimber.
 */

// ============================================================================
// Types
// ============================================================================

export interface MonitorDecision {
  /** Whether the action is allowed */
  allowed: boolean;
  /** Reason for rejection (if not allowed) */
  reason?: string;
  /** Warning message (if allowed but risky) */
  warning?: string;
  /** Suggested alternative action */
  suggestion?: string;
}

export interface ActionContext {
  /** Tool being called */
  toolName: string;
  /** Tool arguments */
  args: Record<string, unknown>;
  /** Current workspace path */
  workspace: string;
  /** Task ID being worked on */
  taskId: string;
  /** Files that have been modified in this session */
  modifiedFiles: string[];
  /** Turn number in the execution */
  turnNumber: number;
  /** Previous actions taken */
  previousActions: string[];
}

// ============================================================================
// Validation Rules
// ============================================================================

/**
 * Rule: Prevent writing outside workspace
 */
function checkWorkspaceBounds(ctx: ActionContext): MonitorDecision | null {
  if (ctx.toolName === "write_file" || ctx.toolName === "edit_file") {
    const path = ctx.args.path as string || ctx.args.file_path as string;
    if (path) {
      // Absolute path outside workspace
      if (path.startsWith("/") && !path.startsWith(ctx.workspace) && !path.startsWith("/app")) {
        return {
          allowed: false,
          reason: `Cannot write outside workspace: ${path}`,
          suggestion: `Use relative path or /app/ prefix`,
        };
      }
      // Parent directory traversal
      if (path.includes("..")) {
        return {
          allowed: false,
          reason: `Path traversal not allowed: ${path}`,
          suggestion: `Use direct path without ..`,
        };
      }
    }
  }
  return null;
}

/**
 * Rule: Prevent dangerous shell commands
 */
function checkDangerousCommands(ctx: ActionContext): MonitorDecision | null {
  if (ctx.toolName === "run_command") {
    const cmd = ctx.args.command as string || "";
    const dangerous = [
      { pattern: /rm\s+-rf?\s+\/(?!app)/, reason: "Cannot delete system directories" },
      { pattern: /rm\s+-rf?\s+\*/, reason: "Cannot delete all files" },
      { pattern: /chmod\s+777/, reason: "Overly permissive chmod" },
      { pattern: /curl.*\|\s*(?:ba)?sh/, reason: "Cannot pipe curl to shell" },
      { pattern: /wget.*\|\s*(?:ba)?sh/, reason: "Cannot pipe wget to shell" },
      { pattern: /sudo/, reason: "Cannot use sudo" },
      { pattern: /\bdd\s+.*of=\/dev/, reason: "Cannot write to block devices" },
    ];

    for (const { pattern, reason } of dangerous) {
      if (pattern.test(cmd)) {
        return {
          allowed: false,
          reason,
          suggestion: "Use a safer alternative",
        };
      }
    }
  }
  return null;
}

/**
 * Rule: Warn about overwriting solution files without testing first
 *
 * NOTE: Task-specific solution file mappings have been removed.
 * This rule now uses a generic heuristic: warn if overwriting any file
 * that was previously modified without testing.
 */
function checkTestBeforeSubmit(ctx: ActionContext): MonitorDecision | null {
  if (ctx.toolName === "write_file") {
    const path = ctx.args.path as string || ctx.args.file_path as string || "";

    // Generic check: warn if overwriting a previously modified file without testing
    if (ctx.turnNumber > 1 && ctx.modifiedFiles.includes(path)) {
      // Check if they've tested before rewriting
      const hasTestedRecently = ctx.previousActions.some(
        (a) => a.includes("verify_progress") || a.includes("test") || a.includes("pytest")
      );

      if (!hasTestedRecently) {
        return {
          allowed: true, // Allow but warn
          warning: `Overwriting ${path} without testing. Consider using verify_progress first.`,
        };
      }
    }
  }
  return null;
}

/**
 * Rule: Prevent infinite loops (same action repeated)
 */
function checkRepetition(ctx: ActionContext): MonitorDecision | null {
  if (ctx.previousActions.length >= 3) {
    const last3 = ctx.previousActions.slice(-3);
    const actionSignature = `${ctx.toolName}:${JSON.stringify(ctx.args)}`;

    // Check if this exact action was done recently
    const repeats = last3.filter((a) => a === actionSignature).length;
    if (repeats >= 2) {
      return {
        allowed: false,
        reason: `Action repeated ${repeats + 1} times: ${ctx.toolName}`,
        suggestion: "Try a different approach or use verify_progress to see current state",
      };
    }
  }
  return null;
}

// ============================================================================
// GUARDRAIL: NO TASK-SPECIFIC HARDCODING
//
// All task-specific validation rules have been removed:
// - checkRegexLogSpecific
// - checkPathTracingSpecific
// - checkDnaAssemblySpecific
//
// This file must NEVER contain:
// - Task IDs (e.g., "regex-log", "path-tracing")
// - Task-specific patterns (e.g., IPv4 format, date format)
// - Task-specific hints (e.g., "use lookahead for IPv4")
// - Task-specific file paths (e.g., "/app/regex.txt")
//
// All knowledge must come from:
// 1. The task description (passed as parameter)
// 2. General process knowledge (TDD, iteration)
//
// If you're tempted to add task-specific code, you're defeating the thesis:
// "Architecture beats model size"
// ============================================================================

// ============================================================================
// Main Monitor
// ============================================================================

/**
 * All validation rules in priority order
 *
 * NOTE: All task-specific validation rules have been removed.
 */
const VALIDATION_RULES = [
  checkWorkspaceBounds,
  checkDangerousCommands,
  checkRepetition,
  checkTestBeforeSubmit,
];

/**
 * Monitor an action before execution.
 *
 * @param ctx Action context with tool, args, and execution state
 * @returns Decision on whether to allow the action
 */
export function monitorAction(ctx: ActionContext): MonitorDecision {
  const warnings: string[] = [];

  for (const rule of VALIDATION_RULES) {
    const decision = rule(ctx);

    if (decision) {
      if (!decision.allowed) {
        // Immediate rejection
        return decision;
      }

      if (decision.warning) {
        warnings.push(decision.warning);
      }
    }
  }

  // All rules passed
  const result: MonitorDecision = { allowed: true };
  if (warnings.length > 0) {
    result.warning = warnings.join("; ");
  }
  return result;
}

/**
 * Create an action signature for tracking
 */
export function createActionSignature(toolName: string, args: Record<string, unknown>): string {
  // Create a normalized signature for comparison
  const normalizedArgs: Record<string, unknown> = {};

  for (const [key, value] of Object.entries(args)) {
    if (key === "content" || key === "new_string") {
      // For content, just use a hash/length to detect same writes
      normalizedArgs[key] = `content:${String(value).length}`;
    } else {
      normalizedArgs[key] = value;
    }
  }

  return `${toolName}:${JSON.stringify(normalizedArgs)}`;
}

/**
 * Check if an action is the same as a previous one
 */
export function isSameAction(
  action1: { toolName: string; args: Record<string, unknown> },
  action2: { toolName: string; args: Record<string, unknown> }
): boolean {
  return createActionSignature(action1.toolName, action1.args) ===
    createActionSignature(action2.toolName, action2.args);
}

File: crates/hillclimber/src/monitor.rs
//! HillClimber Monitor Module
//!
//! Validates actions BEFORE execution to catch obvious mistakes.
//! Rule-based, no LLM needed for fast validation.
//!
//! Part of the MAP (Modular Agentic Planner) architecture.

use crate::types::{ActionContext, MonitorDecision};
use regex::Regex;

// ============================================================================
// GUARDRAIL: NO TASK-SPECIFIC HARDCODING
//
// This file must NEVER contain:
// - Task IDs (e.g., "regex-log", "path-tracing")
// - Task-specific patterns (e.g., IPv4 format, date format)
// - Task-specific hints (e.g., "use lookahead for IPv4")
// - Task-specific file paths (e.g., "/app/regex.txt")
//
// All knowledge must come from:
// 1. The task description (passed as parameter)
// 2. General process knowledge (TDD, iteration)
// ============================================================================

// ============================================================================
// Validation Rules
// ============================================================================

/// Rule: Prevent writing outside workspace.
fn check_workspace_bounds(ctx: &ActionContext) -> Option<MonitorDecision> {
    if ctx.tool_name == "write_file" || ctx.tool_name == "edit_file" {
        let path = ctx
            .args
            .get("path")
            .or_else(|| ctx.args.get("file_path"))
            .and_then(|v| v.as_str());

        if let Some(path) = path {
            // Absolute path outside workspace
            if path.starts_with('/') && !path.starts_with(&ctx.workspace.to_string_lossy().as_ref()) && !path.starts_with("/app") {
                return Some(MonitorDecision::deny_with_suggestion(
                    format!("Cannot write outside workspace: {}", path),
                    "Use relative path or /app/ prefix",
                ));
            }
            // Parent directory traversal
            if path.contains("..") {
                return Some(MonitorDecision::deny_with_suggestion(
                    format!("Path traversal not allowed: {}", path),
                    "Use direct path without ..",
                ));
            }
        }
    }
    None
}

/// Rule: Prevent dangerous shell commands.
fn check_dangerous_commands(ctx: &ActionContext) -> Option<MonitorDecision> {
    if ctx.tool_name == "run_command" {
        let cmd = ctx
            .args
            .get("command")
            .and_then(|v| v.as_str())
            .unwrap_or("");

        // Note: Rust regex doesn't support lookahead (?!), so we use explicit checks

        // Check for rm -rf on dangerous system paths (not /app)
        let rm_rf_re = Regex::new(r"rm\s+-rf?\s+(/\w+)").ok();
        if let Some(re) = &rm_rf_re {
            if let Some(caps) = re.captures(cmd) {
                let path = &caps[1];
                // Allow /app, reject other system paths
                if path != "/app" && !path.starts_with("/app/") {
                    return Some(MonitorDecision::deny_with_suggestion(
                        "Cannot delete system directories".to_string(),
                        "Use a safer alternative",
                    ));
                }
            }
        }

        // Simple patterns without lookahead
        let dangerous_patterns: Vec<(&str, &str)> = vec![
            (r"rm\s+-rf?\s+\*", "Cannot delete all files"),
            (r"chmod\s+777", "Overly permissive chmod"),
            (r"curl.*\|\s*(ba)?sh", "Cannot pipe curl to shell"),
            (r"wget.*\|\s*(ba)?sh", "Cannot pipe wget to shell"),
            (r"\bsudo\b", "Cannot use sudo"),
            (r"\bdd\s+.*of=/dev", "Cannot write to block devices"),
        ];

        for (pattern, reason) in dangerous_patterns {
            if let Ok(re) = Regex::new(pattern) {
                if re.is_match(cmd) {
                    return Some(MonitorDecision::deny_with_suggestion(
                        reason.to_string(),
                        "Use a safer alternative",
                    ));
                }
            }
        }
    }
    None
}

/// Rule: Warn about overwriting files without testing first.
fn check_test_before_submit(ctx: &ActionContext) -> Option<MonitorDecision> {
    if ctx.tool_name == "write_file" {
        let path = ctx
            .args
            .get("path")
            .or_else(|| ctx.args.get("file_path"))
            .and_then(|v| v.as_str())
            .unwrap_or("");

        // Generic check: warn if overwriting a previously modified file without testing
        if ctx.turn_number > 1 && ctx.modified_files.contains(&path.to_string()) {
            // Check if they've tested before rewriting
            let has_tested_recently = ctx.previous_actions.iter().any(|a| {
                a.contains("verify_progress") || a.contains("test") || a.contains("pytest")
            });

            if !has_tested_recently {
                return Some(MonitorDecision::allow_with_warning(format!(
                    "Overwriting {} without testing. Consider using verify_progress first.",
                    path
                )));
            }
        }
    }
    None
}

/// Rule: Prevent infinite loops (same action repeated).
fn check_repetition(ctx: &ActionContext) -> Option<MonitorDecision> {
    if ctx.previous_actions.len() >= 3 {
        let last_3: Vec<&String> = ctx.previous_actions.iter().rev().take(3).collect();
        let action_signature = create_action_signature_internal(&ctx.tool_name, &ctx.args);

        // Check if this exact action was done recently
        let repeats = last_3.iter().filter(|a| a.as_str() == action_signature).count();
        if repeats >= 2 {
            return Some(MonitorDecision::deny_with_suggestion(
                format!("Action repeated {} times: {}", repeats + 1, ctx.tool_name),
                "Try a different approach or use verify_progress to see current state",
            ));
        }
    }
    None
}

// ============================================================================
// Main Monitor
// ============================================================================

/// All validation rules in priority order.
type ValidationRule = fn(&ActionContext) -> Option<MonitorDecision>;

const VALIDATION_RULES: &[ValidationRule] = &[
    check_workspace_bounds,
    check_dangerous_commands,
    check_repetition,
    check_test_before_submit,
];

/// Monitor an action before execution.
///
/// # Arguments
///
/// * `ctx` - Action context with tool, args, and execution state
///
/// # Returns
///
/// Decision on whether to allow the action
pub fn monitor_action(ctx: &ActionContext) -> MonitorDecision {
    let mut warnings: Vec<String> = Vec::new();

    for rule in VALIDATION_RULES {
        if let Some(decision) = rule(ctx) {
            if !decision.allowed {
                // Immediate rejection
                return decision;
            }

            if let Some(warning) = decision.warning {
                warnings.push(warning);
            }
        }
    }

    // All rules passed
    if warnings.is_empty() {
        MonitorDecision::allow()
    } else {
        MonitorDecision::allow_with_warning(warnings.join("; "))
    }
}

/// Create an action signature for tracking.
pub fn create_action_signature(tool_name: &str, args: &serde_json::Value) -> String {
    create_action_signature_internal(tool_name, args)
}

fn create_action_signature_internal(tool_name: &str, args: &serde_json::Value) -> String {
    // Create a normalized signature for comparison
    let mut normalized_args = serde_json::Map::new();

    if let Some(obj) = args.as_object() {
        for (key, value) in obj {
            if key == "content" || key == "new_string" {
                // For content, just use length to detect same writes
                let len = value.as_str().map(|s| s.len()).unwrap_or(0);
                normalized_args.insert(key.clone(), serde_json::json!(format!("content:{}", len)));
            } else {
                normalized_args.insert(key.clone(), value.clone());
            }
        }
    }

    format!(
        "{}:{}",
        tool_name,
        serde_json::to_string(&normalized_args).unwrap_or_default()
    )
}

/// Check if two actions are the same.
pub fn is_same_action(
    action1_tool: &str,
    action1_args: &serde_json::Value,
    action2_tool: &str,
    action2_args: &serde_json::Value,
) -> bool {
    create_action_signature(action1_tool, action1_args)
        == create_action_signature(action2_tool, action2_args)
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::path::PathBuf;

    fn create_test_context(
        tool_name: &str,
        args: serde_json::Value,
    ) -> ActionContext {
        ActionContext {
            tool_name: tool_name.to_string(),
            args,
            workspace: PathBuf::from("/app/workspace"),
            task_id: "test-task".to_string(),
            modified_files: vec![],
            turn_number: 1,
            previous_actions: vec![],
        }
    }

    #[test]
    fn test_workspace_bounds() {
        // Writing to /app/ should be allowed
        let ctx = create_test_context(
            "write_file",
            serde_json::json!({"path": "/app/solution.txt", "content": "test"}),
        );
        let decision = monitor_action(&ctx);
        assert!(decision.allowed);

        // Writing outside workspace should be rejected
        let ctx = create_test_context(
            "write_file",
            serde_json::json!({"path": "/etc/passwd", "content": "test"}),
        );
        let decision = monitor_action(&ctx);
        assert!(!decision.allowed);

        // Path traversal should be rejected
        let ctx = create_test_context(
            "write_file",
            serde_json::json!({"path": "/app/../etc/passwd", "content": "test"}),
        );
        let decision = monitor_action(&ctx);
        assert!(!decision.allowed);
    }

    #[test]
    fn test_dangerous_commands() {
        // Safe command should be allowed
        let ctx = create_test_context(
            "run_command",
            serde_json::json!({"command": "ls -la"}),
        );
        let decision = monitor_action(&ctx);
        assert!(decision.allowed);

        // rm -rf /etc should be rejected
        let ctx = create_test_context(
            "run_command",
            serde_json::json!({"command": "rm -rf /etc"}),
        );
        let decision = monitor_action(&ctx);
        assert!(!decision.allowed);

        // rm -rf /app should be allowed (workspace)
        let ctx = create_test_context(
            "run_command",
            serde_json::json!({"command": "rm -rf /app/temp"}),
        );
        let decision = monitor_action(&ctx);
        assert!(decision.allowed);

        // sudo should be rejected
        let ctx = create_test_context(
            "run_command",
            serde_json::json!({"command": "sudo cat /etc/shadow"}),
        );
        let decision = monitor_action(&ctx);
        assert!(!decision.allowed);
    }

    #[test]
    fn test_repetition_detection() {
        // Create signature with same args as the action (including content)
        let signature = create_action_signature("write_file", &serde_json::json!({"path": "/app/test.txt", "content": "test"}));

        let mut ctx = create_test_context(
            "write_file",
            serde_json::json!({"path": "/app/test.txt", "content": "test"}),
        );
        ctx.previous_actions = vec![
            "other_action".to_string(),
            signature.clone(),
            signature.clone(),
        ];

        let decision = monitor_action(&ctx);
        assert!(!decision.allowed);
        assert!(decision.reason.unwrap().contains("repeated"));
    }

    #[test]
    fn test_action_signature() {
        let sig1 = create_action_signature(
            "write_file",
            &serde_json::json!({"path": "/app/test.txt", "content": "hello"}),
        );
        let sig2 = create_action_signature(
            "write_file",
            &serde_json::json!({"path": "/app/test.txt", "content": "world"}),
        );
        let sig3 = create_action_signature(
            "write_file",
            &serde_json::json!({"path": "/app/other.txt", "content": "hello"}),
        );

        // Same path, same content length -> same signature
        assert_eq!(sig1, sig2);
        // Different path -> different signature
        assert_ne!(sig1, sig3);
    }
}
